File: QID-1332-SFSinterpolMaturity/SFSinterpolMaturity.R

# clear variables and close windows
rm(list = ls(all = TRUE))
graphics.off()

# install and load packages
libraries = c("fOptions")
lapply(libraries, function(x) if (!(x %in% installed.packages())) {
install.packages(x)
})
lapply(libraries, library, quietly = TRUE, character.only = TRUE)

# parameter settings
DAX    = 4617.07                       # Closing price at time t_0.
strike = 4600                          # Strike price
irate  = 0.021                         # Interest rate
mvec   = c(0.2109, 0.4602, 0.7095)     # Vector of option maturities
cvec   = c(119.4, 194.3, 256.9)        # Vector of call prices

# Calculate implied volatilities
impvol = matrix(, , )
for (i in 1:3) impvol[i] = GBSVolatility(cvec[i], "c", DAX, strike, mvec[i], irate, 
    b = irate)
IntpInV = matrix(, , )

# Interpolation in prices
interpolatedprice = (mvec[2] - mvec[1]) * (cvec[3] - cvec[1])/(mvec[3] - mvec[1]) + 
    cvec[1]
IntpInP = interpolatedprice

# Interpolation in volatility
interpvol = (mvec[2] - mvec[1]) * (impvol[3] - impvol[1])/(mvec[3] - mvec[1]) + 
    impvol[1]

# Calculation of option price by BS model with implied volatility
d1 = (log(DAX/strike) + (irate + interpvol^2/2) * mvec[2])/(interpvol * sqrt(mvec[2]))
d2 = d1 - interpvol * sqrt(mvec[2])
interpolatedprice = DAX * pnorm(d1) - strike * exp(-irate * mvec[2]) * pnorm(d2)

IntpInV = interpolatedprice

# Deviation from true price
diff = c(abs(cvec[2] - IntpInP), abs(cvec[2] - IntpInV))

print("Implied volatilities")
print("  C1     C2     C3")
print(round(impvol, 4))

print("Interpolated volatility for C2")
print(round(interpvol, 4))

print("Interpolated prices")
print("by price")
print(IntpInP)
print("by volatility")
print(IntpInV)

print("Deviation from true price")
print("by price")
print(diff[1])
print("by volatility")
print(round(diff[2], 4))

File: QID-1332-SFSinterpolMaturity/SFSinterpolMaturity.m
% ---------------------------------------------------------------------
% Book:        SFS
% ---------------------------------------------------------------------
% Quantlet:    SFSinterpolMaturity
% ---------------------------------------------------------------------
% Description: SFSinterpolMaturity uses linear interpolation of option
%              prices C1 and C3 and implied volatilities in order to
%              approximate the price of another option C2. All options have
%              the same strike, but vary in maturity with tau1<tau2<tau3.
%              Corresponds to exercise 17.1 in SFS.
% ---------------------------------------------------------------------
% Usage:       -
% ---------------------------------------------------------------------
% Inputs:      mvec   - vector of maturities
%              cvec   - vector of call prices
%              strike - strike price of option
%              irate  - Interest rate
%              DAX    - Closing price of underlying at time t_0.
% ---------------------------------------------------------------------
% Output:     impvol    - implied volatilities
%             interpvol - interpolated volatility
%             IntpInP   - interpolated price by prices
%             IntpInV   - interpolated price by volatility
%             diff      - Deviation from true price
% ---------------------------------------------------------------------
% Example:    -
% ---------------------------------------------------------------------
% Author:     Szymon Borak 20090731
% ---------------------------------------------------------------------

clear
clc
close all

DAX    = 4617.07; %Closing price at time t_0.
strike = 4600; %Strike price
irate  = 0.0210; %Interest rate

mvec =  [ 0.2109 0.4602 0.7095]; %Vector of option maturities
cvec = [ 119.4 194.3 256.9 ];    %Vector of call prices

impvol = blsimpv(DAX,strike,irate,mvec,cvec);

method = 'linear';

IntpInP = [];
IntpInV = [];

i = 2;

% interpolation in prices
interpolatedprice = interp1(mvec([i-1,i+1]),cvec([i-1,i+1]),mvec(i),method);
IntpInP           = [IntpInP interpolatedprice];

% interpolation in volatility
interpvol         = interp1(mvec([i-1,i+1]),impvol([i-1,i+1]),mvec(i),method);
interpolatedprice = blsprice(DAX,strike,irate,mvec(i),interpvol);
IntpInV           = [IntpInV interpolatedprice];

%Deviation from true price
diff = [abs(cvec(i)-IntpInP), abs(cvec(i)-IntpInV)];

disp('Implied volatilities')
disp('      C1        C2        C3')
disp(impvol)

disp('Interpolated volatility for C2')
disp(interpvol)

disp('Interpolated prices')
disp(' by price  by volatility')
disp([IntpInP, IntpInV])

disp('Deviation from true price')
disp('   by price  by volatility')
disp(diff)